### PR TITLE
InteractionParameters

### PR DESCRIPTION
**Pull request by chriadam**

### CODE REVIEW

**Comment by chriadam** [comment]
@dcaliste This one implements the API (and the plumbing for the flows), to allow the sort of thing we discussed on IRC a couple of weeks back.  Of course, I haven't yet integrated the password agent plugin which Andrew wrote, so this won't be too useful at the moment ;-)  You can test it with the in-app flow, of course.

**Comment by chriadam** [comment]
I still have to implement the analogous functionality for the Crypto API (i.e. get a passphrase from the user then use a key derivation function to transform that into a key, in the GenerateStoredKeyRequest flow).  I will do that tomorrow.

**Comment by dcaliste** [comment]
Wonderful, I'm going to use it in the pinentry and report before middle of next week. Thank you.

**Comment by chriadam** [comment]
Ok, the symmetric key derivation function support has been added, along with the flows to support user-input as the KDF input data.  I had to refactor a bunch of things, and completely change the binary layout of the Key type in the process, so you'll need to delete your database to test this.

I still need to test that the user-input flow works properly, as I've only tested the non-interactive input data case so far.

**Comment by chriadam** [comment]
I've now tested the user input flows, and they work correctly.  I've also updated the documentation.  I think this is ready for and will merge it unless people raise issues in the next day or so, as the API changes need to go in so that plugin PRs can be rebased etc.

File: daemon/SecretsImpl/secretsrequestprocessor.cpp [hunk 1]
```diff
@@ -1232,6 +1285,96 @@ Daemon::ApiImpl::RequestProcessor::setCollectionSecret(
...
+}
+
+Result
+Daemon::ApiImpl::RequestProcessor::setCollectionSecretGetAuthenticationKey(
+        pid_t callerPid,
+        quint64 requestId,
+        const Secret &secret,
+        SecretManager::UserInteractionMode userInteractionMode,
+        const QString &interactionServiceAddress,
+        bool collectionUsesDeviceLockKey,
```
[the remark this answers is not in the document]
Comment (chriadam) [reply daemon/SecretsImpl/secretsrequestprocessor.cpp]: This is impl detail, we can clean it up later if needed.  Patches welcome.

File: daemon/SecretsImpl/secrets.cpp [hunk 2]
```diff
@@ -473,14 +482,20 @@ void Daemon::ApiImpl::SecretsRequestQueue::handlePendingRequest(
...
                     ? request->inParams.takeFirst().value<Secret>()
                     : Secret();
+            InteractionParameters uiParams = request->inParams.size()
+                    ? request->inParams.takeFirst().value<InteractionParameters>()
+                    : InteractionParameters();
             SecretManager::UserInteractionMode userInteractionMode = request->inParams.size()
                     ? request->inParams.takeFirst().value<SecretManager::UserInteractionMode>()
                     : SecretManager::PreventInteraction;
-            QString interactionServiceAddress = request->inParams.size() ? request->inParams.takeFirst().value<QString>() : QString();
+            QString interactionServiceAddress = request->inParams.size()
```
[the remark this answers is not in the document]
Comment (chriadam) [reply daemon/SecretsImpl/secrets.cpp]: It's probably not optimal.  We should clean it up, if so, but it doesn't have to happen now.  Patches welcome.

**Comment by chriadam** [comment]
It's difficult to split these changes up, meaningfully (a refactor touches everything, and adding some API functionality requires everything (client API, daemon plumbing, plugin implementation, and unit test) because they can't be developed in isolation.  Sure, I could retrospectively go back and split up the work into 4 separate commits (client API, daemon plumbing, plugin implementation, and unit tests) after I've finished, and that might make it easier to review.  But it would also take me 3 extra hours, and that's time I don't have right now...

**Comment by chriadam** [comment]
I've now added the asymmetric cipher key pair generation stuff.

**Pull request merged**